src/index.ts

```diff
@@ -0,0 +1,2721 @@
+    } else if (this.layer === 2) {
+      return FC(Math.sign(this.mag), 1, Math.abs(this.mag) + 0.36221568869946325); //log10(log10(e))
+    } else {
+      return FC(Math.sign(this.mag), this.layer - 1, Math.abs(this.mag));
+    }
+  }
+
+  public logarithm(base: DecimalSource): Decimal {
+    return this.log(base);
+  }
+
+  public pow(value: DecimalSource): Decimal {
+    const decimal = D(value);
+    const a = this;
+    const b = decimal;
+
+    //special case: if a is 0, then return 0
+    if (a.sign === 0) {
+      return a;
+    }
+    //special case: if a is 1, then return 1
+    if (a.sign === 1 && a.layer === 0 && a.mag === 1) {
+      return a;
+    }
+    //special case: if b is 0, then return 1
+    if (b.sign === 0) {
+      return FC_NN(1, 0, 1);
+    }
+    //special case: if b is 1, then return a
+    if (b.sign === 1 && b.layer === 0 && b.mag === 1) {
+      return a;
+    }
+
+    const result = a.absLog10().mul(b).pow10();
+
+    if (this.sign === -1 && b.toNumber() % 2 === 1) {
+      return result.neg();
+    }
+
+    return result;
+  }
+
+  public pow10(): Decimal {
+    /*
+    There are four cases we need to consider:
+    1) positive sign, positive mag (e15, ee15): +1 layer (e.g. 10^15 becomes e15, 10^e15 becomes ee15)
+    2) negative sign, positive mag (-e15, -ee15): +1 layer but sign and mag sign are flipped (e.g. 10^-15 becomes e-15, 10^-e15 becomes ee-15)
+    3) positive sign, negative mag (e-15, ee-15): layer 0 case would have been handled in the Math.pow check, so just return 1
+    4) negative sign, negative mag (-e-15, -ee-15): layer 0 case would have been handled in the Math.pow check, so just return 1
+    */
+
+    if (!Number.isFinite(this.layer) || !Number.isFinite(this.mag)) {
+      return Decimal.dNaN;
+    }
+
+    let a = this;
+
+    //handle layer 0 case - if no precision is lost just use Math.pow, else promote one layer
+    if (a.layer === 0) {
+      const newmag = Math.pow(10, a.sign * a.mag);
+      if (Number.isFinite(newmag) && Math.abs(newmag) > 0.1) {
+        return FC(1, 0, newmag);
+      } else {
+        if (a.sign === 0) {
+          return Decimal.dOne;
+        } else {
+          a = FC_NN(a.sign, a.layer + 1, Math.log10(a.mag)) as this;
+        }
+      }
+    }
+
+    //handle all 4 layer 1+ cases individually
+    if (a.sign > 0 && a.mag > 0) {
+      return FC(a.sign, a.layer + 1, a.mag);
+    }
+    if (a.sign < 0 && a.mag > 0) {
+      return FC(-a.sign, a.layer + 1, -a.mag);
+    }
+    //both the negative mag cases are identical: one +/- rounding error
+    return Decimal.dOne;
+  }
+
+  public pow_base(value: DecimalSource): Decimal {
+    return D(value).pow(this);
+  }
+
+  public root(value: DecimalSource): Decimal {
+    const decimal = D(value);
+    return this.pow(decimal.recip());
+  }
+
+  public factorial(): Decimal {
+    if (this.mag < 0) {
+      return this.toNumber().add(1).gamma(); // FIXME: This is an invalid operation, number doesn't have an add
```

Comment:
`this.toNumber()` cannot have `add` called on it since it's a regular `number` type.

```suggestion
      return this.add(1).gamma();
```